src/slap/ext/application/venv.py

```diff
@@ -20,22 +20,38 @@
     "bash": """
     function slap() {
       local ORIGINAL=$(which slap)
-      if ! [ $? = 0 ]; then
+      if [[ $? != 0 ]]; then
         >&2 echo "error: command 'slap' does not exist"
         return 127
       fi
-      if [ "$1" == "venv" ] && [[ "$2" =~ -[gc]*a[gc]* ]]; then
+      if [[ "$1" == "venv" && "$2" =~ -[gc]*a[gc]* ]]; then
         eval "$(SLAP_SHADOW=true "$ORIGINAL" "$@")"
       else
         "$ORIGINAL" "$@"
       fi
       return $?
     }
-  """,
+    """,
+    "zsh": """
+    function slap() {
+      local ORIGINAL=$(command which slap)
+      if [[ $? != 0 ]]; then
+        >&2 echo "error: command 'slap' does not exist"
+        return 127
+      fi
+      if [[ "$1" == "venv" && "$2" =~ -[gc]*a[gc]* ]]; then
+        eval "$(SLAP_SHADOW=true "$ORIGINAL" "$@")"
+      else
+        "$ORIGINAL" "$@"
+      fi
+      return $?
+    }
+    """,
 }
 
 USER_INIT_SCRIPTS = {
     "bash": 'which slap >/dev/null && eval "$(SLAP_SHADOW=true slap venv -i bash)"',
+    "zsh": 'which slap >/dev/null && eval "$(SLAP_SHADOW=true slap venv -i zsh)"',
```

Comment:
i really don't know anything about zsh, but extrapolating from the other zsh code piece above, wouldn't this need to be `command which slap` instead?

Reply:
well honestly, i'm not entirely sure i fully understand this either, but bear with me.

the relevant difference between bash and zsh is that function definitions shadow commands when calling `which` in zsh, which doesn't happen in bash.

zsh:
```zsh
» vim () { echo "I'm not actually vim"; }
» vim
I'm not actually vim
» which vim
vim () {
	echo "I'm not actually vim"
}
```

bash:
```bash
$ vim () { echo "I'm not actually vim"; }
$ vim
I'm not actually vim
$ which vim
/opt/homebrew/bin/vim
```

by prefixing `which slap` with `command` in [line 38](https://github.com/NiklasRosenstein/slap/pull/74/files#diff-4eef4a4275c7a683f7efe24f74f283fdf9a0ae1d745d17d52728076391ebcf6dR37), we are letting zsh know to ignore local function definitions and use the global definition, which is the path to slap (as opposed to the shadow function); if we don't do this, zsh gives us:

```zsh
» which slap
slap () {
	local ORIGINAL=$(command which slap)
	if [[ $? != 0 ]]
	then
		echo "error: command 'slap' does not exist" >&2
		return 127
	fi
	if [[ "$1" == "venv" && "$2" =~ -[gc]*a[gc]* ]]
	then
		eval "$(SLAP_SHADOW=true "$ORIGINAL" "$@")"
	else
		"$ORIGINAL" "$@"
	fi
	return $?
}
```

and that doesn't work because this replaces `$ORIGINAL` with the shadow function, resulting in an infinite recursion (which zsh gracefully handles by telling us that "file is too long", lol).

**however**, when the init script gets called (when loading .zshrc), the shadow function isn't in place yet. if we remove the init script call from .zshrc, we get:

```zsh
» which slap
/Users/florian.eich/.local/bin/slap
```

which is what the init script gets, and which is the intended behaviour... i think? :D

Reply:
I think this makes sense, thanks for the in-depth exaplanation 😉

Reply:
what do i need to do to get this merged then? :)

Reply:
Rebasing I believe 😄

Reply:
yeah that seems fair enough alright :P

Reply:
and done ✅